### PR TITLE
resource/aws_transfer_user: add home_directory_type and home_directory_mappings arguments

### LINKED ISSUE #11281: aws_transfer_user should provide home-directory-mappings option.
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS CLI https://docs.aws.amazon.com/cli/latest/reference/transfer/create-user.html provides` home-directory-type` and `home-directory-mappings` to control logical mapping between S3 and SFTP. Terraform AWS provider does not support the feature.

### New or Affected Resource(s)

* aws_transfer_user

### Potential Terraform Configuration

```hcl
resource "aws_transfer_user" "sftp_user_dbs" {
  # other options
  # ...
  home_directory_type = "LOGICAL"
  home_directory_mapping = {
    entry = "your-personal-report.pdf"
    target = "/bucket3/customized-reports/${transfer:UserName}.pdf" 
  }
}
```

### References

* https://docs.aws.amazon.com/cli/latest/reference/transfer/create-user.html

### PR DESCRIPTION
**Pull request by quinyx-tjeerd**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Original PR:  
https://github.com/terraform-providers/terraform-provider-aws/pull/11483

I resolved the conflict

Closes #11281
Relates #11632

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
Add home_directory_type and home_directory_mappings arguments to aws_transfer_user resource
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSTransferUser_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAWSTransferUser_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSTransferUser_basic
=== PAUSE TestAccAWSTransferUser_basic
=== RUN   TestAccAWSTransferUser_modifyWithOptions
=== PAUSE TestAccAWSTransferUser_modifyWithOptions
=== RUN   TestAccAWSTransferUser_disappears
=== PAUSE TestAccAWSTransferUser_disappears
=== RUN   TestAccAWSTransferUser_UserName_Validation
=== PAUSE TestAccAWSTransferUser_UserName_Validation
=== RUN   TestAccAWSTransferUser_homeDirectoryMappings
=== PAUSE TestAccAWSTransferUser_homeDirectoryMappings
=== CONT  TestAccAWSTransferUser_basic
=== CONT  TestAccAWSTransferUser_UserName_Validation
=== CONT  TestAccAWSTransferUser_disappears
=== CONT  TestAccAWSTransferUser_homeDirectoryMappings
=== CONT  TestAccAWSTransferUser_modifyWithOptions
--- PASS: TestAccAWSTransferUser_UserName_Validation (15.55s)
--- PASS: TestAccAWSTransferUser_disappears (45.52s)
--- PASS: TestAccAWSTransferUser_basic (49.50s)
--- PASS: TestAccAWSTransferUser_homeDirectoryMappings (73.62s)
--- PASS: TestAccAWSTransferUser_modifyWithOptions (112.28s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	113.609s
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/flatmap	0.421s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags	0.671s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/naming	0.890s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/apigatewayv2/waiter	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/batch/equivalency	0.829s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/eks/token	0.236s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/guardduty/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/iam/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/kinesisanalytics/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/kms/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/neptune/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/rds/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/secretsmanager/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/servicediscovery/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/workspaces/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/awsproviderlint	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/helper/awsprovidertype/keyvaluetags	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes	1.587s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSAT001	2.180s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSR001	0.613s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSR002	1.231s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/fmtsprintfcallexpr	0.462s [no tests to run]
...
```

### CODE REVIEW

**Comment by sallaben** [comment]
Thanks for the PR! Does this functionality differ from the functionality in #12599?

**Comment by quinyx-tjeerd** [comment]
This functionality is a subset of the pr you mention, which could (and possibly should) have been separated so it could add value sooner. I could revise the code so it matches the mentioned pr?

**Comment by rprieto** [comment]
Is it possible to merge this PR? It's indeed covered by 12599 - however that one implements several features and seems to be blocked on an unrelated topic, waiting for another PR itself (VPC endpoint). Without this functionality we must fall back to CloudFormation to manage AWS Transfer users, especially as directory mappings are not available in the AWS Console. Thanks.

**Comment by dturnbu** [comment]
Who is this "AWS Provider" guy and why hasn't he reviewed this PR? I've been waiting for this for weeks now and it looks like others have been waiting even longer.

How is it this simple, yet valuable fix hasn't made it into one of the several releases since this PRs creation?

**Comment by quinyx-tjeerd** [comment]
In case it helps you out, I worked around this issue by:
adding AWS cli to the image where terraform runs, and using local-exec.
this might or might not be an option for you, depending on where your terraform runs.

```
resource "aws_transfer_user" "sftp-users" {
  server_id = aws_transfer_server.server.id
  role      = aws_iam_role.sftp-user.arn
  user_name = "username"

  # This bit can be replaced once 'home-directory-type' and 'home-directory-mappings' are available in terraform aws provider
  provisioner "local-exec" {
    command = "aws transfer update-user --server-id $SERVER --user-name $USER --region $REGION --home-directory-type $HOME_DIR_TYPE --home-directory-mappings $HOME_DIR_MAPPINGS"

    environment = {
      SERVER            = aws_transfer_server.server.id
      USER              = "username"
      REGION            = var.region
      HOME_DIR_TYPE     = "LOGICAL"
      HOME_DIR_MAPPINGS = format("Entry=/,Target=/%s/%s", aws_s3_bucket.sftp-bucket.id, "username")
    }
  }
}
```

**Comment by dturnbu** [comment]
> In case it helps you out, I worked around this issue by:
> adding AWS cli to the image where terraform runs, and using local-exec.
> this might or might not be an option for you, depending on where your terraform runs.
> 
> ```
> resource "aws_transfer_user" "sftp-users" {
>   server_id = aws_transfer_server.server.id
>   role      = aws_iam_role.sftp-user.arn
>   user_name = "username"
> 
>   # This bit can be replaced once 'home-directory-type' and 'home-directory-mappings' are available in terraform aws provider
>   provisioner "local-exec" {
>     command = "aws transfer update-user --server-id $SERVER --user-name $USER --region $REGION --home-directory-type $HOME_DIR_TYPE --home-directory-mappings $HOME_DIR_MAPPINGS"
> 
>     environment = {
>       SERVER            = aws_transfer_server.server.id
>       USER              = "username"
>       REGION            = var.region
>       HOME_DIR_TYPE     = "LOGICAL"
>       HOME_DIR_MAPPINGS = format("Entry=/,Target=/%s/%s", aws_s3_bucket.sftp-bucket.id, "username")
>     }
>   }
> }
> ```

Thank you very much, I'll give it a shot! I'm trying to run this from Terraform Cloud, hopefully their executors allow local-exec operations.....

**Comment by quinyx-tjeerd** [comment]
> > In case it helps you out, I worked around this issue by:
> > adding AWS cli to the image where terraform runs, and using local-exec.
> > this might or might not be an option for you, depending on where your terraform runs.
> > ```
> > resource "aws_transfer_user" "sftp-users" {
> >   server_id = aws_transfer_server.server.id
> >   role      = aws_iam_role.sftp-user.arn
> >   user_name = "username"
> > 
> >   # This bit can be replaced once 'home-directory-type' and 'home-directory-mappings' are available in terraform aws provider
> >   provisioner "local-exec" {
> >     command = "aws transfer update-user --server-id $SERVER --user-name $USER --region $REGION --home-directory-type $HOME_DIR_TYPE --home-directory-mappings $HOME_DIR_MAPPINGS"
> > 
> >     environment = {
> >       SERVER            = aws_transfer_server.server.id
> >       USER              = "username"
> >       REGION            = var.region
> >       HOME_DIR_TYPE     = "LOGICAL"
> >       HOME_DIR_MAPPINGS = format("Entry=/,Target=/%s/%s", aws_s3_bucket.sftp-bucket.id, "username")
> >     }
> >   }
> > }
> > ```
> 
> Thank you very much, I'll give it a shot! I'm trying to run this from Terraform Cloud, hopefully their executors allow local-exec operations.....

I used Terraform Cloud before at a previous company, and local-exec was not the problem but i'd image you don't have AWS cli available in the images they run on 😅  
and back then I haven't looked into the functionality / flexibility of providing your own images.

File: aws/resource_aws_transfer_user_test.go [hunk 1]
```diff
@@ -650,3 +659,122 @@ resource "aws_transfer_user" "foo" {
...
+			"Action": "sts:AssumeRole"
+		}
+	]
+}
+EOF
+}
+
+resource "aws_iam_role_policy" "foo" {
+  name = "tf-test-transfer-user-iam-policy-%s"
+  role = "${aws_iam_role.foo.id}"
```
Comment (bflad) [inline aws/resource_aws_transfer_user_test.go:687]: Nit: We are now using Terraform 0.12+ syntax in test configurations, will adjust on merge. 👍

**Review by bflad — APPROVED** [review]
This looks good, thanks for the updates here, @quinyx-tjeerd 🚀 

Output from acceptance testing:

```
--- PASS: TestAccAWSTransferUser_basic (20.00s)
--- PASS: TestAccAWSTransferUser_disappears (16.77s)
--- PASS: TestAccAWSTransferUser_homeDirectoryMappings (30.78s)
--- PASS: TestAccAWSTransferUser_modifyWithOptions (43.38s)
--- PASS: TestAccAWSTransferUser_UserName_Validation (8.93s)
```

**Pull request merged**